pkg/windows/calico.go

```diff
@@ -52,147 +137,232 @@ const (
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create token for service account (%s/%s)", CalicoSystemNamespace, calicoNode)
+	}
+
+	calicoKubeConfig.Token = token.Status.Token
 
-	return cfg, nil
+	return &calicoKubeConfig, nil
 }
 
 // Start starts the CNI services on the Windows node.
-func (c *Calico) Start(ctx context.Context, config *CNIConfig) error {
+func (c *Calico) Start(ctx context.Context) error {
 	for {
-		if err := startCalico(ctx, config.CalicoConfig); err != nil {
+		if err := startCalico(ctx, c.CNICfg); err != nil {
 			continue
 		}
 		break
 	}
-	go startFelix(ctx, config.CalicoConfig)
+	go startFelix(ctx, c.CNICfg)
 
 	return nil
 }
 
-// createNodeNameFile creates the node name file required by the CNI.
-func createNodeNameFile(config *CNIConfig, dataDir string) error {
-	return util2.WriteFile(config.CalicoConfig.KubeConfig.Path, config.NodeConfig.AgentConfig.NodeName)
-}
+func (c *Calico) generateCalicoNetworks() error {
+	if err := deleteAllNetworksOnNodeRestart(); err != nil {
+		return err
+	}
 
-// createKubeConfig creates the kube config required by the CNI.
-func createKubeConfig(config *CNIConfig, dataDir string) error {
-	kubeTemplate, err := parseTemplateFromConfig(calicoKubeConfigTemplate, config)
+	mgmt, err := createHnsNetwork(c.CNICfg.Mode, os.Getenv("VXLAN_ADAPTER"))
 	if err != nil {
 		return err
 	}
-	return util2.WriteFile(filepath.Join("c:\\", dataDir, "agent", CalicoKubeConfigName), kubeTemplate)
+
+	if c.CNICfg.Platform == "ec2" || c.CNICfg.Platform == "gce" {
+		logrus.Debugf("recreating metadata route because platform is: %s", c.CNICfg.Platform)
+		if err := setMetaDataServerRoute(mgmt); err != nil {
+			return err
+		}
+	}
+	if c.CNICfg.Mode == "windows-bgp" {
+		// Not supported yet, this does work.
+		_ = wapi.StopService("RemoteAccess")
+		_ = wapi.StartService("RemoteAccess")
+	}
+	return nil
 }
 
-// createCNIConfig creates the CNI config for the CNI.
-func createCNIConfig(config *CNIConfig) error {
-	parsedTemplate, err := parseTemplateFromConfig(calicoConfigTemplate, config)
+// overrideCalicoConfigByHelm overrides the default values set for calico if a Chart exists
+func (c *Calico) overrideCalicoConfigByHelm(restConfig *rest.Config) error {
+	hc, err := helm.NewFactoryFromConfig(restConfig)
 	if err != nil {
 		return err
 	}
-	return util2.WriteFile(filepath.Join(config.NodeConfig.AgentConfig.CNIConfDir, CalicoConfigName), parsedTemplate)
-}
 
-// getDefaultConfig sets the default configuration.
-func getDefaultConfig(config *CNIConfig, dataDir string, nodeConfig *config.Node) error {
-	platformType, err := getPlatformType()
+	cniChartConfig, err := hc.Helm().V1().HelmChartConfig().Get(metav1.NamespaceSystem, CalicoChart, metav1.GetOptions{})
 	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to check for %s HelmChartConfig", CalicoChart)
+	}
+	if cniChartConfig == nil {
```

Comment:
Should we log here to know why this may have returned nil?

Reply:
good idea, let me add a log when being in debug mode